### ISSUE TITLE
Update from 3.6 to 4.1 requires package reinstall

### ISSUE THREAD

**Issue by RichardD2** [aspnet/RoslynCodeDomProvider#151]
I have now updated two separate MVC5 projects from v3.6 to v4.1. In both instances, the package update claims to have succeeded. But after the update, attempting to run the site produces an error:

```
Could not find a part of the path '[Root of the site]\roslyn\csc.exe'.
   at System.Web.Compilation.BuildManager.ReportTopLevelCompilationException()
   at System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled()
   at System.Web.Compilation.BuildManager.CallAppInitializeMethod()
   at System.Web.Hosting.HostingEnvironment.Initialize(ApplicationManager appManager, IApplicationHost appHost, IConfigMapPathFactory configMapPathFactory, HostingEnvironmentParameters hostingParameters, PolicyLevel policyLevel, Exception appDomainCreationException)

Could not find a part of the path '[Root of the site]\roslyn\csc.exe'.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at Microsoft.CodeDom.Providers.DotNetCompilerPlatform.Compiler.get_CompilerName()
   at Microsoft.CodeDom.Providers.DotNetCompilerPlatform.Compiler.FromFileBatch(CompilerParameters options, String[] fileNames)
   at Microsoft.CodeDom.Providers.DotNetCompilerPlatform.Compiler.CompileAssemblyFromFileBatch(CompilerParameters options, String[] fileNames)
   at System.Web.Compilation.AssemblyBuilder.Compile()
   at System.Web.Compilation.BuildProvidersCompiler.PerformBuild()
   at System.Web.Compilation.ApplicationBuildProvider.GetGlobalAsaxBuildResult(Boolean isPrecompiledApp)
   at System.Web.Compilation.BuildManager.CompileGlobalAsax()
   at System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled()
```

If I then reinstall the NuGet package:

```
update-package Microsoft.CodeDom.Providers.DotNetCompilerPlatform -reinstall
```

the site starts working again.

The `web.config` file changes after the first update: the `system.codeDom/compilers/compiler` elements change from `Version=3.6.0.0` to `Version=4.1.0.0`, and the C# one changes from `/langversion:7.3` to `/langversion:default`.

After the reinstall, the only difference is in the project file - from:

```
<Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
  <PropertyGroup>
    <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
  </PropertyGroup>
  <Error Condition="!Exists('..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.4.1.0\build\net472\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.4.1.0\build\net472\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets'))" />
  <Error Condition="!Exists('..\packages\Microsoft.Net.Compilers.Toolset.4.5.0\build\Microsoft.Net.Compilers.Toolset.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.Net.Compilers.Toolset.4.5.0\build\Microsoft.Net.Compilers.Toolset.props'))" />
</Target>
```

to:

```
<Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
  <PropertyGroup>
    <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
  </PropertyGroup>
  <Error Condition="!Exists('..\packages\Microsoft.Net.Compilers.Toolset.4.5.0\build\Microsoft.Net.Compilers.Toolset.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.Net.Compilers.Toolset.4.5.0\build\Microsoft.Net.Compilers.Toolset.props'))" />
  <Error Condition="!Exists('..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.4.1.0\build\net472\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.4.1.0\build\net472\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets'))" />
</Target>
```

**Comment by StephenMolloy** [comment]
My first question might be dumb... but before running the MVC app, did you re-build the application after updating the package? (The 4.X releases of this package are msbuild-centric and use a .targets file to pull roslyn tools into the app build at build time. The previous 2.X and 3.X releases used nuget install.ps1 scripts to push roslyn tools into the application at package install time rather than app build time. This is the direction nuget and msbuild/dotnet are moving, so the 4.X releases try to align with this new direction rather than cling to the old ways.) Assuming you are building after updating the package...

I don't think the issue is with the project file. Based on the exception text, I see that it is missing 'bin/' in the path, but you say this is an MVC application. So I'd expect '/bin' to be a part of the path. One possible cause would be that you've got a special roslyn compiler location configured with either the `ROSLYN_COMPILER_LOCATION` environment variable or the `aspnet:RoslynCompilerLocation` appSetting in web.config. It's worth crossing those off the list, but I don't think those are the likely culprit.

Also, since you mention a project file, I assume you're not using a project-less "Website" that has a [known issue](#149) with the 4.X series of this provider.

Instead I think the [default "runtime" tool-path probing](https://github.com/aspnet/RoslynCodeDomProvider/blob/main/src/Microsoft.CodeDom.Providers.DotNetCompilerPlatform/Util/CompilationUtil.cs#L202) is getting tripped up. Probably because the 'bin/roslyn' folder hasn't been created when you start your app. However, I'm unable to reproduce your issue, so it's hard to speculate further on what might be happening.

If you have a small sample app in the state that reproduces the issue that you could provide, that would be quite helpful in diagnosing the root cause. You could also do some build log exploring of your own to see if you can determine why the 'bin\roslyn' folder isn't being created during build. In the past, I've had success using binlog to figure out why things run or get skipped in complex builds. The task that copies Roslyn binaries into the app is named [CopyRoslynCompilerFilesToOutputDirectory](https://github.com/aspnet/RoslynCodeDomProvider/blob/main/src/Packages/Microsoft.CodeDom.Providers.DotNetCompilerPlatform/build/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets#L41). Additionally, the task that determines _where_ to copy the roslyn tools to is named [LocateRoslynToolsDestinationFolder](https://github.com/aspnet/RoslynCodeDomProvider/blob/main/src/Packages/Microsoft.CodeDom.Providers.DotNetCompilerPlatform/build/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.targets#L29) - perhaps the destination is being miscalculated in your case.

**Comment by RichardD2** [comment]
Thankfully, I managed to get rid of the last of our project-less website projects last year. 😊

I've not managed to reproduce this error recently. I'll let you know if it comes up again.

I did have one project a couple of weeks ago where I had to manually delete the folder from the `packages` folder to get it to update from 3.6 to 4.1; before I did that, even a reinstall of the package was installing an older version of the compilers. But I don't know if that's related to this issue.
